registry/aws-sns-topic/src/index.js

```
@@ -0,0 +1,210 @@
+      {},
+      topicAttributes
+    ),
+    { deliveryStatusAttributes: flatDeliveryStatusAttributes }
+  )
+}
+
+const updateTopicAttributes = async ({ topicAttributes, topicArn }) =>
+  Promise.all(
+    map((topicAttribute) => {
+      const value = first(values(topicAttribute))
+      const params = {
+        TopicArn: topicArn,
+        AttributeName: capitalize(first(keys(topicAttribute))),
+        AttributeValue: typeof value !== 'string' ? JSON.stringify(value) : value
+      }
+      return sns.setTopicAttributes(params).promise()
+    }, topicAttributes)
+  )
+
+const updateDeliveryStatusAttributes = async ({ deliveryStatusAttributes, topicArn }) =>
+  // run update requests sequentially because setTopicAttributes
+  // fails to update when rate exceeds https://github.com/serverless/components/issues/174#issuecomment-390463523
+  resolveInSequence(
+    map(
+      (topicAttribute) => () => {
+        const value = first(values(topicAttribute))
+        const params = {
+          TopicArn: topicArn,
+          AttributeName: capitalize(first(keys(topicAttribute))),
+          AttributeValue: typeof value !== 'string' ? JSON.stringify(value) : value
+        }
+        return sns.setTopicAttributes(params).promise()
+      },
+      deliveryStatusAttributes
+    )
+  )
+
+const removeSNSTopic = async ({ topicArn }) =>
+  sns
+    .deleteTopic({
+      TopicArn: topicArn
+    })
+    .promise()
+
+const remove = async (inputs, context) => {
+  context.log(`Removing SNS topic: '${context.state.name}'`)
+  await removeSNSTopic(context.state)
+  context.saveState({})
+  return {
+    arn: null
+  }
+}
+
+const deploy = async (inputs, context) => {
+  const { state } = context
+  let newState
+  context.setOutputs({
+    arn: null
+  })
+
+  if (!state.name && inputs.name) {
+    // if no name stored to state, create a new topic
+    newState = await createSNSTopic(inputs, context)
```

Similar to the above idempotent scenario for remove, we should aim for the same behavior here. It is possible for the state to become desynchronized with the reality of the infrastructure. So, this should be thought of as more of a set operation than a create/update. (note that we have NOT gone back and updated our component to fix this problem in deploy yet). 

If the topic already exists, but we think that it doesn't, swallow the error and update the attributes instead.
If the topic does exist but we think that it doesn't, swallow the error, create the topic and then update the attributes.

---

@brianneisler same thing with this. I deployed a new topic; then I set the state to an empty object in the `state.json` so that the `serviceId` and `instanceId` remained same. After that, I deployed again, and the response was identical with clean deployment.

---

I probably need to remove the topic and create a new if topic attributes changes because there is no method for removing those attributes. But I'll get into that later.

---

@brianneisler I think I've manually tested all the cases that I can come up with. The SNS API is not too picky with non-excisting resources (it doesn't throw errors...), but there was one role/policy issue https://github.com/serverless/components/pull/215/commits/c35957eaea9b5e36fb8d1566008964e7d4182377.